PublicAPI/Microsoft.ApplicationInsights.dll/net45/PublicAPI.Unshipped.txt

```diff
@@ -33,5 +33,237 @@ Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.IHeartbeatPro
 Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.IHeartbeatPropertyManager.IsHeartbeatEnabled.get -> bool
 Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.IHeartbeatPropertyManager.IsHeartbeatEnabled.set -> void
 Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing.IHeartbeatPropertyManager.SetHeartbeatProperty(string propertyName, string propertyValue = null, bool? isHealthy = null) -> bool
-Microsoft.ApplicationInsights.DataContracts.PageViewTelemetry.Id.get -> string
-Microsoft.ApplicationInsights.DataContracts.PageViewTelemetry.Id.set -> void
\ No newline at end of file
+Microsoft.ApplicationInsights.Metric
+Microsoft.ApplicationInsights.Metric.GetAllSeries() -> System.Collections.Generic.IReadOnlyList<System.Collections.Generic.KeyValuePair<string[], Microsoft.ApplicationInsights.Metrics.MetricSeries>>
+Microsoft.ApplicationInsights.Metric.GetDimensionValues(int dimensionNumber) -> System.Collections.Generic.IReadOnlyCollection<string>
+Microsoft.ApplicationInsights.Metric.Identifier.get -> Microsoft.ApplicationInsights.Metrics.MetricIdentifier
+Microsoft.ApplicationInsights.Metric.SeriesCount.get -> int
+Microsoft.ApplicationInsights.Metric.TrackValue(double metricValue) -> void
+Microsoft.ApplicationInsights.Metric.TrackValue(object metricValue) -> void
+Microsoft.ApplicationInsights.Metric.TryGetDataSeries(out Microsoft.ApplicationInsights.Metrics.MetricSeries series) -> bool
+Microsoft.ApplicationInsights.Metric.TryGetDataSeries(out Microsoft.ApplicationInsights.Metrics.MetricSeries series, bool createIfNotExists, params string[] dimensionValues) -> bool
+Microsoft.ApplicationInsights.Metric.TryGetDataSeries(out Microsoft.ApplicationInsights.Metrics.MetricSeries series, string dimension1Value) -> bool
+Microsoft.ApplicationInsights.Metric.TryGetDataSeries(out Microsoft.ApplicationInsights.Metrics.MetricSeries series, string dimension1Value, string dimension2Value) -> bool
```

Comment:
why do we need `Try` alternative? Why not limit it to unconditional method? What customer will do when it returned false?

Reply:
Thank you for this question.
The Try-pattern is employed here in-line with the Framework Design Guidelines, because the method for tracking a value may fail even if the user specifies valid parameters under non-trivial conditions:

A key goal of the metric APIs is to ensure that a service cannot fail due to mistakes in tracing and telemetry collection (nothing is absolute, so we strive for "unlikely to fail"). This includes preventing memory starvation due to unintentional combinatorial explosion in the number if time series.
For that a `MetricConfiguration` has options for `SeriesCountLimit` and `ValuesPerDimensionLimit`. If these limits are reached, but a `TryTrackValue(..)` call would result in the creation of a new time series, the call will return `false` and the value will not be tracked.

There are different approaches for handling dimension capping, so rather than prescribing the customer a specific approach, we make it easy to detect this case and to react to the circumstance. Please let me know if you would like me to discuss some of these approaches.

Notably, we do not offer exception-throwing `TrackValue(..)` equivalents to all `TryTrackValue(..)` methods. This is because appropriate usage of such `TrackValue(..)` methods would require them to always be used in a try-catch block. This is not appropriate for tracing and telemetry collection APIs, which should be easy to incorporate in existing code and must not break existing code.

Please let me know if this answers your question.
If you require adjustments to these APIs, please request *precisely* the changes you would like to see so that I can make them.
Thank you.